File: bench/ref/sort.c
/* hand-written C reference for bench/sort.jqd (task 75): the SAME merge
 * sort over heap-allocated cons nodes — the claim compares the language's
 * list discipline against C practicing the same discipline, not against a
 * flat-array qsort (which measures the representation, not the compiler). */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct node {
  int64_t v;
  struct node *next;
} node;

static node *cons(int64_t v, node *next) {
  node *n = malloc(sizeof(node));
  n->v = v;
  n->next = next;
  return n;
}

/* split into alternating halves, sort each, merge — list.sort's shape */
static node *merge(node *a, node *b) {
  node head = { 0, NULL };
  node *t = &head;
  while (a && b) {
    if (a->v <= b->v) { t->next = a; a = a->next; }
    else { t->next = b; b = b->next; }
    t = t->next;
  }
  t->next = a ? a : b;
  return head.next;
}

static node *msort(node *xs) {
  if (!xs || !xs->next) return xs;
  node *a = NULL, *b = NULL;
  while (xs) {
    node *n = xs->next;
    xs->next = a; a = xs;
    xs = n;
    if (xs) { n = xs->next; xs->next = b; b = xs; xs = n; }
  }
  return merge(msort(a), msort(b));
}

int main(void) {
  node *xs = NULL;
  /* list.range is half-open: exactly 200000 nodes, like the .jqd program */
  for (int64_t i = 0; i < 200000; i++) xs = cons(i, xs);
  xs = msort(xs);
  int64_t len = 0;
  for (node *p = xs; p; p = p->next) len++;
  printf("%lld\n", (long long)len);
  return 0;
}

File: bench/ref/sort.py
# Python twin of bench/sort.jqd with the same heap discipline: merge sort
# over heap-allocated nodes, not the built-in sort over a flat list.
class Node:
    __slots__ = ('v', 'next')
    def __init__(self, v, nxt):
        self.v = v
        self.next = nxt
def merge(a, b):
    head = Node(0, None)
    t = head
    while a and b:
        if a.v <= b.v:
            t.next = a
            a = a.next
        else:
            t.next = b
            b = b.next
        t = t.next
    t.next = a if a else b
    return head.next
def msort(xs):
    if not xs or not xs.next:
        return xs
    a = b = None
    while xs:
        n = xs.next
        xs.next = a
        a = xs
        xs = n
        if xs:
            n = xs.next
            xs.next = b
            b = xs
            xs = n
    return merge(msort(a), msort(b))
xs = None
for i in range(200000):
    xs = Node(i, xs)
xs = msort(xs)
n = 0
while xs:
    n += 1
    xs = xs.next
print(n)
